### ISSUE TITLE
Migrate to using Babel-Acorn

### ISSUE THREAD

**Issue by nmn** [facebook/flow#640]
Please use the AST generated by the Babel-parser instead of maintaining your own parser that generates a SpiderMonkey AST. This will free you up from the work of adding new language features to parser as Babel is already way ahead on that work.

Of course, this will not get rid of the amount of work needed to support new features, as the work to analyze the AST, do the flow analysis and find error would still be a huge task. Also, the migration would probably be a big task as well. I'm guessing the SpiderMonkey AST and the AST that babel generates are different in big and small ways. But I think it's still worth it in the long term.

Some other reasons this is a good idea:

1) Sebastian, the maintainer of Babel has indicated that he wants to build much better linting tools by integrating with Flow. Using the same AST in both tools will be mutually beneficial to both tools
1a) Sebastian is also soon going to be Facebook employ, so Facebook will still have same level of control over their projects

2) Babel already does some error checking on it's own. For example, `let` and `const` block-scoping is already done in Babel. Babel also throws errors if you try to re-assign a `const`. By teaming up with Babel, flow could worry about implementing fewer checks.

There are some possible concerns as well. The biggest one being that Flow is written entirely in Ocaml, including the parser. Babel is written in Javascript instead. So it's possible that transferring the AST from one language to the other may be a bit of a challenge. However, ASTs are generally serializable, so it should not be too big of a problem.

**Comment by abritinthebay** [comment]
:+1:  to this, could not agree more.

**Comment by gabelevi** [comment]
There are really two parts to this:

1) On which AST should Flow operate
2) Which parser should Flow use to generate that AST

= Which AST to use =

We started off targeting the spider monkey AST, since that was a popular standard at the time. Recently, it's been superseded by the [estree](https://github.com/estree/estree) spec. I haven't invested too heavily in the parser since estree came into existence, so some work is probably needed, but the intention is to target estree. My guess is that we're pretty close at the moment. Babel intends to target estree as well.

= Which parser to use =

We wrote our own parser in OCaml because it compiles to native bytecode and is way faster than JavaScript. We had considered using an existing parser written in JavaScript and serializing the AST but performance is very important to us (we need to work on massive repositories) and operationally compiling in a parser is easier.

= One parser to rule them all =

It is a problem that we have to building things into multiple parsers. Ideally, we would just have one parser that we could use everywhere. The Flow parser does compile to JavaScript. Someday when we find the time, I think it would be awesome to spend some time optimizing the js_of_ocaml version of the Flow parser, and plugging that into Babel. But I have no plans to work on that any time soon.

**Comment by mroch** [comment]
Yeah, I think the perf and logistics overhead of communicating between JS (node?) and ocaml is enough to kill this for now... sorry. :/

We definitely feel this pain though, because Facebook has to make sure the flow parser, babel *and* [jstransform](https://github.com/facebook/jstransform) parse things before we can use anything new. we're moving away from jstransform because of this, and consolidating to one would be even better.
